infra/src/main/kotlin/fi/fta/geoviite/infra/split/SplitValidation.kt

```diff
@@ -89,3 +89,13 @@ internal fun validateTargetGeometry(
             }
     }
 }
+
+internal fun validateSplitSourceLocationTrack(
+    source: SplitSource,
+    locationTrack: LocationTrack,
+): PublishValidationError? {
+    return if (source.locationTrackId == locationTrack.id && locationTrack.exists) PublishValidationError(
```

Comment:
Ei mikään must-juttu, mutta näissä validoinneissa voi myös käyttää validate-funkkaria publicationvalidation.kt:sta. Se on toiminnaltaan vastaava kuin kotlinin requiret ym, mutta ei heitä vaan palauttaa julkaisuerroin tai nullin. Semanttisena erona että se on "validoi että X" eikä "jos !X" kuten tällä rakentella -> iffin ehto pitää kääntää kuvaamaan validia casea:
```
return validate(source.locationTrackId != locationTrack.id || !locationTrack.exists) {
  "$VALIDATION_SPLIT.source-not-deleted"
}
```

Tässä yksittäisessä se vain hiukan siistii poistamalla iffin. Isompi hyöty on silloin jos validoi monta asiaa, jolloin voi käyttää listOfNotNull:ia tuottamaan kaikki virheet yhtenä nippuna.

Reply:
Anekdoottina mulla menee ainakin aivot usein solmuun kun näitä näkee, kun joutuu päässään miettimään mitä tuo validate() tekee & kääntämään tuon "validate()n" kuitenkin iffiksi.

Reply:
Njoo katsantokannasta kiinni ehkä. Minusta tuo check/require/validate/assert/jne syntaksi on nimenomaa se selkeämpi koska siinä varmistetaan että onhan asia ehjä ja sellainen kun sen kuuluu olla `validate(<is ok>)`. Negaatiota tässä on minusta se että `if(<is not ok>) throw...` Tässä casessa toki itse tarkastettava ehto sattuu menemään noin päin että good case on paljon negaatiota itse logiikassa, mutta tietty tuolle voisi olla aisaparina `validateNot` kuten asserteilla ja muillakin vastaavilla yleensä on.

Reply:
Tulkinnallisesti välittömänä negaationa tarkoitin, että virhe palautetaan JOS ehto EI päde, eli näkyvissä olevan boolean-logiikan joutuu päässään kääntämään, jotta tietää milloin virhe palautetaan. 

Iffittelyssä taas if(<is not ok>) logiikka on välittömästi nähtävissä eli käänteispäättelyä ei edellytetä eikä työmuistia tarvita yhtä paljon. Sinänsä siis validatedNot voisi toimia (joskin sekin on negaatio).

Vähän sellainen fiilis, että custom-assertointien käyttö normaalien iffien sijaan on vähän turha epäsuoruus. Palautusarvosta tai heitosta olisi jo kuitenkin varsin selvää, että kyseessä on validaatio tai muu tarkistus. Tarkempia validaatiofunkkareita (validateSwitch() jne) pitäisin kuitenkin ihan mielekkäinä.

Tässä tietysti myös varsinaiseen ehtoon päätyy vielä negaatiota itsessään, joka sitovat aivoja entisestään, etenkin mitä enemmän ANDia ja ORia löytyy, mutta se on granulaarisempi pointti kuin mitä tässä hain.

Reply:
Njoo toki suoran `if(x) throw ..` tyylisen kohdalla sen funktiointi on aika pieni ero mutta itse pidän sitä silti selkeämpänä. En nimittäin aukipura tuota validate-abstraktiota koodia lukiessa sen enempää kuin mitään muutakaan funktiota, vaan tämän funktion kontekstia katsoessa vilkaisen vain että tässä validoidaan x ja mennään sitten eteenpäin. 

Se oikea etu noista toki alkaa tulla vasta sitten kun niitä on enemmän. Muutama peräkkäinen validate näkyy jo selvemmin että tässä validoidaan kaikki nuo jutut sen sijaan että olisi monta iffiä jotka pitää katsoa tarkemmin ettei joku niistä haaroista tee jotain muuta. Lisäksi virheet palauttava validointi menee kivasti lyhyemmäksi tyylillä `val errors = lisOfNotNull(validate(x), validate(y), validate(z))` sen sijaan että monessa iffissä pitäisi summailla virheitä johonkin listalle. 

Tässä ei toki eroa kun kyseessä vain yksi, paitsi ehkä se että jokaiselle tällaiselle validoitavalle asialle ei tarvitse tehdä omaa nimettyä funktiota, varsinkin jos tarkastetaan vain yksi asia ja sen ehto on simppeli.

Noh, makuasiaa varmasti ja mitä on tottunut lukemaan. Itse pidän tuosta kotlin-tyylisestä assertiosta ja se istuu minusta tähän validate-konseptiin myös nätisti.